added Polynomial cutoff function as descibed by Peterson and Khorshidi

EricMusa opened the pull request:
*Includes derivatives. 
Tested with basic script --> compared gamma=2. polynomial function against cosine function (similar forms, as depicted in fig. 5 of ([Peterson & Khorshidi](https://www.sciencedirect.com/science/article/pii/S0010465516301266?via%3Dihub)) --> similar values and derivative values, everything checks out

Next step is to look into adding control flow in the symmetry function code to use either the cosine cutoff function or the polynomial function for calculations

mshuaibii commented:
Thanks for the PR @EricMusa. Lets merge this once we get the control flow in place and try training some models with it being used. If you'd like to work build off this that would be great.

EricMusa reviewed (commented):
**Modifications for choosing Polynomial vs Cosine cutoff function made:**
 - I've added the "cutoff_params" dict to the "descriptor_setup" tuple of descriptor settings.
 - "cutoff_params" by default will be {'cutoff_func': 'cosine'} but can also be of the form {'cutoff_func': 'polynomial', 'gamma': <float>}. 
 - This get's passed to the Gaussian descriptor object which, when using _libsymf for calculations, now calls either "calculate_sf_cos" or "calculate_sf_poly" (or their noderiv complements) depending on the passed cutoff function. 
 - I've modified "example.py" to create two models - each with a descriptor set using one of the two cutoff functions - and then print the accuracies of the two models. 
 - The script runs fine and the results are very similar.

EricMusa reviewed (commented):
couple of bug fixed and more testing:
 - modified Gaussian.py get_descriptor_setup_hash to include info about the cutoff function
    - no longer overwriting same Gs with diff cutoff functions

mshuaibii commented:
Awesome - can you please merge master into this. I've included some CI tests so lets make sure this looks fine with that. If you haven't already done so installing the `pre-commit` hooks (`black` specifically) will take care of formatting/structure to be consistent with the rest of the repo. That's usually the thing people fail on in the tests, just an FYI.

mshuaibii commented:
> couple of bug fixed and more testing:
> 
> * modified Gaussian.py get_descriptor_setup_hash to include info about the cutoff function
>   
>   * no longer overwriting same Gs with diff cutoff functions

Awesome - can you please merge master into this. I've included some CI tests so lets make sure this looks fine with that. If you haven't already done so installing the `pre-commit` hooks (`black` specifically) will take care of formatting/structure to be consistent with the rest of the repo. That's usually the thing people fail on in the tests, just an FYI.

EricMusa reviewed (commented):
`black` pre-commit hook and trim-tailing-whitespace reformatted a lot of the files, also made some various edits to imports to pass `flake8` hook as well.

mshuaibii commented:
Can you run something like this - I had CI builds off for forked-PRs, should be fixed now: `git commit --allow-empty -m "trigger build"`

mshuaibii commented on amptorch/tests/consistency_test.py:
```diff
@@ -121,7 +120,7 @@ def test_energy_force_consistency():
     torch_Cu_bias_2 = torch.FloatTensor(Cu_weights_2[-1, :])
 
-    descriptor_setup = ("gaussian", Gs, elements)
+    descriptor_setup = ("gaussian", Gs, {"cutoff_func": "cosine"}, elements)
```
This descriptor setup is starting to look a little awkward with the nested dictionary. Can we do this differently/cleaner so it's more clear what things are?

```
descriptor_setup  = {
    "descriptor": Gaussian,
    "params": Gs,
    "cutoff_func": ...
    "elements":...
}
```

mshuaibii commented:
Looks good - minor changes to make things more clear/easier to understand.

mshuaibii commented:
@EricMusa FYI you only needed to `trigger build` once, you should be fine to push your commits and builds fine.

EricMusa commented:
Yea, sorry about that, I'm learning as I go haha
Changes:
- reverted example.py back to the upstream/master branch's version (as it was before)
- fixed nested dict in consistency_test.py
- added cutoff_funcs_test.py to test Cosine CF vs Polynomial CF (gamma=2.0)
